sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypes.scala

```diff
@@ -79,3 +78,42 @@ case class CreateStruct(children: Seq[Expression]) extends Expression {
     InternalRow(children.map(_.eval(input)): _*)
   }
 }
+
+/**
+ * Creates a struct with the given field names and values
+ *
+ * @param children Seq(name1, val1, name2, val2, ...)
+ */
+case class CreateNamedStruct(children: Seq[Expression]) extends Expression {
+  assert(children.size % 2 == 0, "NamedStruct expects an even number of arguments.")
```

Comment:
yes, please use `checkInputTypes` here to check `children.size % 2 == 0` and all name expressions are non-null literal string.